### ISSUE TITLE
pre-training speed, data loader and gpu utilization

### ISSUE THREAD

**Issue by meliksahturker** [AnswerDotAI/ModernBERT#236]
We are working towards training a ModernBERT from scratch on our custom dataset for a new language by following the exact same recipe.

However, we are unable to reach the GPU utilization and speed reported in paper, that is, 1.7T tokens trained in 194 hours on 8xH100SXM GPUs.
In fact we only get about 15 to 20 % of this speed.

We also ran training speed experiments on 8xA100SXM and training speed decreased only very slightly, which led us to believe that the bottleneck is the data loader and H100 GPUs are not utilized as they should have been.

In order to reduce the potential overheads that can be introduced by reading the data from disk and tokenizing it on-fly during training, we are looking to pre-tokenize the dataset and load it all to memory, however we were unable to detect any relevant config in the [YAML file here](https://github.com/AnswerDotAI/ModernBERT/blob/pretraining_documentation/yamls/modernbert/modernbert-base-pretrain.yaml).

Therefore, our questions are:
- Does the train script load the whole dataset into memory before starting training? If not, is it possible to enable that and how?
- What is the recipe to pre-tokenize the train dataset and use this in order to alleviate the process of tokenization during training?
- Beyond two suspected reasons and solutions above, what might be the reason(s) behind the slowness in this case? Do you have any recommendations?

**Comment by NohTow** [comment]
Hello,

Indeed, we were using a pre-tokenizer dataset to speed training up.
Those are not detailed in the configs because we did not release the pretraining data.
I don't think I have a boilerplate available right now, but essentially you just have to create a MDS dataset with "input_ids" related to your pre-training data and [it'll be loaded.](https://github.com/AnswerDotAI/ModernBERT/blob/8c57a0f01c12c4953ead53d398a36f81a4ba9e38/src/text_data.py#L500)
Please make sure you are using non-streaming dataset as well, as it uses more memory and is generally slower.

I think people managed to get some data ready by taking inspiration from [this script](https://github.com/AnswerDotAI/ModernBERT/blob/8c57a0f01c12c4953ead53d398a36f81a4ba9e38/src/convert_dataset.py#L247) (make sure to use `token_ids` and not `tokens`).

Else I have this old boilerplate that should be a cool start:
```python
import argparse
import json
import os
import tqdm
from transformers import AutoTokenizer, set_seed
from streaming import StreamingDataset, MDSWriter
import multiprocessing as mp
from functools import partial
import numpy as np
import uuid
import gc
import streaming

MDS_COLS_PRE_TOKENIZED = {
    'input_ids': 'ndarray:uint16',
    # 'attention_mask': 'ndarray:uint16',
    'id': 'str',
    'len': 'int'
}


def get_uuid():
    return str(uuid.uuid4())

def tokenize_batch(tokenizer, batch):
    tokenized = tokenizer(batch, truncation=False, padding=False, return_tensors="np")
    # Convert each sequence individually to uint16
    input_ids = [np.array(seq, dtype=np.uint16) for seq in tokenized['input_ids']]
    # attention_mask = [np.array(seq, dtype=np.uint16) for seq in tokenized['attention_mask']]
    return input_ids

def process_chunk(chunk, tokenizer):
    assert isinstance(chunk, list) and all(isinstance(item, dict) for item in chunk), "Chunk should be a list of dictionaries"
    texts = [item["text"] for item in chunk]
    ids = [item["id"] if "id" in item else get_uuid() for item in chunk]
    if "domain" in chunk[0]:
        domains = [item["domain"] for item in chunk]

    # input_ids, attention_mask = tokenize_batch(tokenizer, texts)
    # assert len(ids) == len(input_ids) == len(attention_mask) == len(texts), f"Length mismatch in chunk. {len(ids)} {len(input_ids)} {len(attention_mask)} {len(texts)}"
    # return [{'id': id, 'input_ids': input_id, 'attention_mask': mask, "len": len(input_id)}
    #         for id, input_id, mask in zip(ids, input_ids, attention_mask)]

    input_ids = tokenize_batch(tokenizer, texts)
    if "domain" in chunk[0]:
        assert len(ids) == len(input_ids) == len(texts), f"Length mismatch in chunk. {len(ids)} {len(input_ids)} {len(texts)}"
        return [{'id': id, 'input_ids': input_id, "len": len(input_id), "domain": domain}
                for id, input_id, domain in zip(ids, input_ids, domains)]
    else:
        assert len(ids) == len(input_ids) == len(texts), f"Length mismatch in chunk. {len(ids)} {len(input_ids)} {len(texts)}"
        return [{'id': id, 'input_ids': input_id, "len": len(input_id)}
                for id, input_id in zip(ids, input_ids)]

def sample_dataset_from_config(args):
    assert os.path.isdir(args.dataset), f"Dataset {args.dataset} does not exist."
    print(f"Using local dataset {args.dataset}...")
    # clean up the shared memory
    streaming.base.util.clean_stale_shared_memory()
    dataset = StreamingDataset(local=args.dataset, shuffle=False, split=None, batch_size=1, keep_zip=False)

    print(f"Using tokenizer model {args.tokenizer}...")
    tokenizer = AutoTokenizer.from_pretrained(args.tokenizer, use_fast=True, add_prefix_space=True)

    output_dir = args.dataset + "-tokenized"

    num_tokens = 0
    num_truncated_tokens = 0
    batch_size = 5000
    if args.has_domains:
        MDS_COLS_PRE_TOKENIZED["domain"] = "str"

    with MDSWriter(out=output_dir, columns=MDS_COLS_PRE_TOKENIZED, compression='zstd') as mds_writer:
        pbar = tqdm.tqdm(total=len(dataset), desc="Processing samples")
        
        for i in range(0, len(dataset), batch_size):
            # Get a single batch of data
            end = min(i + batch_size, len(dataset))
            chunk = [dataset[k] for k in range(i, end)]
            
            if not chunk:
                break
            
            # Process the chunk directly without multiprocessing
            batch_results = process_chunk(chunk, tokenizer)
            
            # Write results immediately
            for item in batch_results:
                mds_writer.write(item)
                num_tokens += item['len']
                num_truncated_tokens += min(1024, item['len'])
                
                # Clear item from memory
                del item
            
            # Clear results from memory
            del batch_results
            gc.collect()
            
            pbar.update(len(chunk))
            
            if args.debug and pbar.n >= 100 * batch_size:
                break

    print(f"Finished writing with a total of {num_tokens} train tokens.")
    # save a json file in the directory with the number of tokens and the number of truncated tokens
    with open(os.path.join(output_dir, "num_tokens.json"), "w") as f:
        f.write(json.dumps({"num_tokens": num_tokens, "num_truncated_tokens": num_truncated_tokens}, indent=2))

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-d", "--dataset", type=str, required=True)
    parser.add_argument("-t", "--tokenizer", type=str, required=True)
    parser.add_argument("--has_domains", action="store_true")
    parser.add_argument("--debug", action="store_true")
    args = parser.parse_args()

    set_seed(123456789)
    sample_dataset_from_config(args)

```

Please note that you'll have to [decompress the dataset](https://github.com/AnswerDotAI/ModernBERT/blob/main/src/data/mds_conversion.py) and also that this do not store attention_masks and will create them during data loading, to save some space (so be careful not to let any padding tokens).
Hope it helps.
(cc @orionw maybe we should add a standard boilerplate for data creation by curating what we have rn)

**Comment by meliksahturker** [comment]
Pre-tokenization indeed allowed achieving the speed reported in the paper. It was sufficient to create a dataset with `id`, `input_ids` and `len` fields.
There are two remarks here:
- Training without pretokenization results in about **30%** of the original speed on H100 SXM GPUs. Since bottleneck is the data loader, the speed does not really change on A100SXM either. Therefore, for optimum GPU utilization, pretokenization of the data is a must.
- Without pretokenization and **chunking** the text into parts that will fit into model's max_len, the text is [truncated ](https://github.com/AnswerDotAI/ModernBERT/blob/8c57a0f01c12c4953ead53d398a36f81a4ba9e38/src/text_data.py#L503)by the data loader and most of the long examples are essentially unused.

**Comment by NohTow** [comment]
Thanks for reporting!

**Comment by ahxxm** [comment]
Should this be documented or fixed someday? I also found that on a single H100 training is fast enough to make the bottleneck perceivable, while it's not obvious on a A10G * 4 setup.

My guess on root cause is such:

> On-disk Dataset -> DataLoader(multiprocess) -> SequencePacker(single thread from main) -> ButterIterable(thread)

This blocks the hot path, because in the main thread(roughly speaking), GPU still has to wait for [_create_batch](https://github.com/AnswerDotAI/ModernBERT/blob/8c57a0f01c12c4953ead53d398a36f81a4ba9e38/src/sequence_packer.py#L189)
* During which, ButterIterable merely preloads text, instead of packed sequence
* Then `NoStreamingDataset` needs to tokenize text [one by one](https://github.com/AnswerDotAI/ModernBERT/blob/8c57a0f01c12c4953ead53d398a36f81a4ba9e38/src/text_data.py#L509) instead of by batch
* And seq packer still need to be fast enough

I don't have a clean/minimal way to make it work like w/o seq packing, but by disabling it my H100 can work at above 600W and at around 230000 tok/s, which translates to `1.7 * 1000000000000 / (230000 * 3600 * 8) ~= 256` hours if we ignore the communication overhead, this should be the benefit of packing on training speed

**Comment by ahxxm** [comment]
I've another 2 problems, does the tokenizer need to add special tokens or not? Because [__getitem__](https://github.com/AnswerDotAI/ModernBERT/blob/8c57a0f01c12c4953ead53d398a36f81a4ba9e38/src/text_data.py#L503) could remove EOS token `[SEP]`. And a minor one, `uint16` limits range to [0, 65535], for multilingual model that has 125000/200000 vocab size, uint32 is necessary

**Comment by orionw** [comment]
> Should this be documented or fixed someday?

Yes, but I think many of us original ModernBERT authors don't have time to add new features. We'd gladly accept a PR or otherwise it does work well with pre-tokenization. 

Re: documentation, since the data wasn't open for ModernBERT there was an effort to reproduce it with open-data. That has more details on [data preparation](https://github.com/JHU-CLSP/ettin-encoder-vs-decoder/blob/main/pretraining/data_processing/README.me). We should add that to this Github's README though.

> And a minor one, uint16 limits range to [0, 65535], for multilingual model that has 125000/200000 vocab size, uint32 is necessary

This is 100% true, but surprisingly if you forget to change this your model will still perform decently well - don't ask me how I know 😅 

[mmBERT](https://arxiv.org/abs/2509.06888) was released recently and used this training repo but made sure to use the larger int size. If you want to continue pre-train a multilingual model I would recommend swapping out the ModernBERT checkpoint for mmBERT. If you are pre-training from scratch, I would recommend making sure to include the `add_prefix_space` in the tokenizer (our biggest mistake in mmBERT and caused worse performance for NER/POS).

> does the tokenizer need to add special tokens or not? Because [getitem](https://github.com/AnswerDotAI/ModernBERT/blob/8c57a0f01c12c4953ead53d398a36f81a4ba9e38/src/text_data.py#L503) could remove EOS token [SEP].

Yes the pre-tokenization should include special tokens. Re: removing EOS token if the sequence length is too long, up to you: for decoders it typically is truncated (do you want your model to learn EOS positions that aren't natural) but for encoders I think it could go either way. We let it truncate as needed, though this was taken care of in the data preparation phase as no input passages were longer than our sequence length anyway due to the pre-tokenization. So that line in `getitem` never gets used.

**Comment by ahxxm** [comment]
I'd surely submit a PR if I can get some clean and reviewable solution!

> your model will still perform decently well - don't ask me how I know

I happened to go through the same.. I trained a cased multilingual tokenizer from scratch for ~30 languages, added post processor template to include CLS and SEP, then due to [an upstream bug](https://github.com/huggingface/transformers/issues/3107) all cased vocab weren't used in pretraining and extension, still, the final model performs decently well in a few classification tasks.

> do you want your model to learn EOS positions that aren't natural

Yeah that makes sense, I guess the EOS/SEP token will still appear in training because of short texts, and for encoding both CLS or mean pooling are not affected. I'll try pre-tokenize to have some numbers and fun. Update: with pre-tokenized dataset and sequence packing, speed bumps from 230000 to 260000 tokens per second!
